perfzero/lib/benchmark.py

```diff
@@ -34,67 +35,95 @@
 class BenchmarkRunner(object):
   """Execute benchmark and report results."""
 
-  def __init__(self, config=None):
+  def __init__(self, config):
+    self.config = config
     project_dir = os.path.abspath(os.path.dirname(os.path.dirname(__file__)))
-    self.workspace_dir = os.path.join(project_dir, 'workspace')
+    self.workspace_dir = os.path.join(project_dir, config.workspace_dir)
     self.site_packages_dir = os.path.join(self.workspace_dir, 'site-packages')
     self.auth_token_path = os.path.join(
         self.workspace_dir, 'auth_tokens/benchmark_upload_gce.json')
     self.output_root_dir = os.path.join(self.workspace_dir, 'output')
-    self.config = config
-    self._setup()
+    self.benchmark_execution_time = {}
 
   def _setup(self):
+    """Download data and checkout git repository."""
+    # Set up the raid array.
+    start_time = time.time()
+    device_utils.create_drive_from_devices('/data',
```

Comment:
I would move device creation / raid creation into the setup because I would also do the download in the setup.  Even if the files exist the download takes time making sure they exist and this slows down someone from running a few tests.  I think setup should be:
-  docker
- data directory
- data for testing

At that point you have everything you need, the big items are take care of and you are ready for testing..

Reply:
Discussed offline. We plan to only keep the docker image build operation in setup.py. Every other setup operation will be performed on demand in benchmark.py. This will not cause observable performance overhead since these setup operation will be skipped by default of benchmark.py founds that e.g. the local_path for GCS data already exists.